File: backend/services/rule_engine.py
from utils.data_cleaner import clean_price, clean_discount
import re

def validate_product(product):
    """
    Enhanced validation for deep compliance fields
    Checks 14+ compliance rules including regulatory requirements
    """
    violations = []
    passed_rules = []
    total_rules = 0
    score = 100

    def add_violation(rule_id, rule_name, message, severity, field=None, penalty=0):
        violations.append({
            "id": rule_id,
            "rule": rule_name,
            "message": message,
            "severity": severity,
            "field": field
        })
        return penalty

    # Extract fields
    price = clean_price(product.get("price"))
    mrp = clean_price(product.get("mrp"))
    discount = clean_discount(product.get("discount"))
    name = product.get("product_name") or product.get("name")
    weight = product.get("weight")
    ingredients = product.get("ingredients", "N/A")
    manufacturer_name = product.get("manufacturer_name", "N/A")
    manufacturer_address = product.get("manufacturer_address", "N/A")
    fssai_number = product.get("fssai_number", "N/A")
    expiry_date = product.get("expiry_date", "N/A")
    
    # OCR Data
    fssai_ocr = product.get("fssai_from_ocr", "N/A")

    # === BASIC VALIDATIONS ===
    
    total_rules += 1
    if not name or str(name).strip() in ("", "N/A"):
        score -= add_violation("VAL001", "Product Name Check", "Missing product name", "CRITICAL", "name", 30)
    else:
        passed_rules.append("VAL001")
    
    total_rules += 1
    if not price or price <= 0:
        score -= add_violation("VAL002", "Price Check", "Missing or invalid price", "HIGH", "price", 25)
    else:
        passed_rules.append("VAL002")
    
    total_rules += 1
    if not weight or weight == "N/A":
        score -= add_violation("VAL003", "Weight Check", "Missing weight/quantity information", "MEDIUM", "weight", 15)
    else:
        passed_rules.append("VAL003")

    if mrp and price and mrp < price:
        total_rules += 1
        score -= add_violation("VAL004", "MRP Compliance", "MRP lower than selling price (illegal)", "CRITICAL", "mrp", 40)
    
    # === REGULATORY COMPLIANCE VALIDATIONS ===
    
    total_rules += 1
    effective_fssai = fssai_ocr if fssai_ocr != "N/A" else fssai_number
    if effective_fssai == "N/A" or not effective_fssai:
        score -= add_violation("REG001", "FSSAI License", "Missing FSSAI license number (required for food products)", "CRITICAL", "fssai_number", 35)
    elif len(str(effective_fssai).strip()) != 14:
        score -= add_violation("REG001", "FSSAI Format", f"Invalid FSSAI number format (must be 14 digits)", "HIGH", "fssai_number", 25)
    else:
        passed_rules.append("REG001")
    
    total_rules += 1
    if manufacturer_name == "N/A" or not manufacturer_name:
        score -= add_violation("REG002", "Manufacturer Name", "Missing manufacturer name (legally required)", "HIGH", "manufacturer_name", 30)
    else:
        passed_rules.append("REG002")
    
    total_rules += 1
    if manufacturer_address == "N/A" or not manufacturer_address:
        score -= add_violation("REG003", "Manufacturer Address", "Missing manufacturer address (legally required)", "MEDIUM", "manufacturer_address", 25)
    else:
        passed_rules.append("REG003")
    
    total_rules += 1
    if ingredients == "N/A" or not ingredients or len(ingredients.strip()) < 5:
        score -= add_violation("REG004", "Ingredients List", "Missing or incomplete ingredient list", "HIGH", "ingredients", 20)
    else:
        passed_rules.append("REG004")
    
    total_rules += 1
    if expiry_date == "N/A" or not expiry_date:
        score -= add_violation("REG005", "Expiry Date", "Missing expiry/best before date", "MEDIUM", "expiry_date", 15)
    else:
        passed_rules.append("REG005")

    # Ensure score doesn't go below 0
    score = max(0, score)

    return {
        "rule_score": score,
        "violations": violations,
        "passed_rules": passed_rules,
        "total_rules": total_rules
    }

File: node_backend/services/ruleEngine.js
/**
 * Rule Engine — equivalent to services/rule_engine.py
 * Validates product compliance against FSSAI and other Indian regulations.
 */
const { cleanPrice, cleanDiscount } = require('../utils/dataCleaner');

function validateProduct(product, deepScrapeAvailable = true) {
  const violations = [];
  const passedRules = [];
  let totalRules = 0;
  let score = 100;

  function addViolation(ruleId, ruleName, message, severity, field = null, penalty = 0) {
    violations.push({ id: ruleId, rule: ruleName, message, severity, field });
    return penalty;
  }

  function addAdvisory(ruleId, ruleName, message, field = null) {
    violations.push({
      id: ruleId,
      rule: ruleName,
      message: message + ' (could not verify — deep scrape unavailable)',
      severity: 'INFO',
      field,
    });
  }

  // Extract fields
  const price = cleanPrice(product.price);
  const mrp = cleanPrice(product.mrp);
  const name = product.product_name || product.name;
  const weight = product.weight;
  const ingredients = product.ingredients || 'N/A';
  const manufacturerName = product.manufacturer_name || 'N/A';
  const manufacturerAddress = product.manufacturer_address || 'N/A';
  const fssaiNumber = product.fssai_number || 'N/A';
  const expiryDate = product.expiry_date || 'N/A';
  const fssaiOcr = product.fssai_from_ocr || 'N/A';

  // === BASIC VALIDATIONS (always checked) ===

  totalRules++;
  if (!name || String(name).trim() === '' || String(name).trim() === 'N/A') {
    score -= addViolation('VAL001', 'Product Name Check', 'Missing product name', 'CRITICAL', 'name', 30);
  } else {
    passedRules.push('VAL001');
  }

  totalRules++;
  if (!price || price <= 0) {
    score -= addViolation('VAL002', 'Price Check', 'Missing or invalid price', 'HIGH', 'price', 25);
  } else {
    passedRules.push('VAL002');
  }

  totalRules++;
  if (!weight || weight === 'N/A') {
    score -= addViolation('VAL003', 'Weight Check', 'Missing weight/quantity information', 'MEDIUM', 'weight', 15);
  } else {
    passedRules.push('VAL003');
  }

  if (mrp && price && mrp < price) {
    totalRules++;
    score -= addViolation('VAL004', 'MRP Compliance', 'MRP lower than selling price (illegal)', 'CRITICAL', 'mrp', 40);
  }

  // === REGULATORY COMPLIANCE VALIDATIONS ===

  totalRules++;
  const effectiveFssai = fssaiOcr !== 'N/A' ? fssaiOcr : fssaiNumber;
  if (effectiveFssai === 'N/A' || !effectiveFssai) {
    if (deepScrapeAvailable) {
      score -= addViolation('REG001', 'FSSAI License', 'Missing FSSAI license number (required for food products)', 'CRITICAL', 'fssai_number', 35);
    } else {
      addAdvisory('REG001', 'FSSAI License', 'FSSAI license number not found on listing', 'fssai_number');
    }
  } else if (String(effectiveFssai).trim().length !== 14) {
    score -= addViolation('REG001', 'FSSAI Format', 'Invalid FSSAI number format (must be 14 digits)', 'HIGH', 'fssai_number', 25);
  } else {
    passedRules.push('REG001');
  }

  totalRules++;
  if (manufacturerName === 'N/A' || !manufacturerName) {
    if (deepScrapeAvailable) {
      score -= addViolation('REG002', 'Manufacturer Name', 'Missing manufacturer name (legally required)', 'HIGH', 'manufacturer_name', 30);
    } else {
      addAdvisory('REG002', 'Manufacturer Name', 'Manufacturer name not found on listing', 'manufacturer_name');
    }
  } else {
    passedRules.push('REG002');
  }

  totalRules++;
  if (manufacturerAddress === 'N/A' || !manufacturerAddress) {
    if (deepScrapeAvailable) {
      score -= addViolation('REG003', 'Manufacturer Address', 'Missing manufacturer address (legally required)', 'MEDIUM', 'manufacturer_address', 25);
    } else {
      addAdvisory('REG003', 'Manufacturer Address', 'Manufacturer address not found on listing', 'manufacturer_address');
    }
  } else {
    passedRules.push('REG003');
  }

  totalRules++;
  if (ingredients === 'N/A' || !ingredients || String(ingredients).trim().length < 5) {
    if (deepScrapeAvailable) {
      score -= addViolation('REG004', 'Ingredients List', 'Missing or incomplete ingredient list', 'HIGH', 'ingredients', 20);
    } else {
      addAdvisory('REG004', 'Ingredients List', 'Ingredient list not found on listing', 'ingredients');
    }
  } else {
    passedRules.push('REG004');
  }

  totalRules++;
  if (expiryDate === 'N/A' || !expiryDate) {
    if (deepScrapeAvailable) {
      score -= addViolation('REG005', 'Expiry Date', 'Missing expiry/best before date', 'MEDIUM', 'expiry_date', 15);
    } else {
      addAdvisory('REG005', 'Expiry Date', 'Expiry date not found on listing', 'expiry_date');
    }
  } else {
    passedRules.push('REG005');
  }

  // Ensure score doesn't go below 0
  score = Math.max(0, score);

  return {
    rule_score: score,
    violations,
    passed_rules: passedRules,
    total_rules: totalRules,
    deep_scrape_available: deepScrapeAvailable,
  };
}

module.exports = { validateProduct };
